src/ChangeStream.php

```diff
@@ -0,0 +1,106 @@
+<?php
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB;
+
+use MongoDB\Driver\Cursor;
+use MongoDB\Driver\Exception\RuntimeException;
+use MongoDB\Exception\ResumeTokenException;
+use IteratorIterator;
+use Iterator;
+
+/**
+ * Iterator for the changeStream command.
+ *
+ * @api
+ * @see \MongoDB\Collection::changeStream()
+ * @see http://docs.mongodb.org/manual/reference/command/changeStream/
+ */
+class ChangeStream implements Iterator
+{
+    private $resumeToken;
+    private $resumeCallable;
+    private $csIt;
+    private $key;
+
+    public function __construct(Cursor $cursor, callable $resumeCallable)
+    {
+        $this->resumeCallable = $resumeCallable;
+        $this->csIt = new IteratorIterator($cursor);
+        $this->csIt->rewind();
+        $this->key = 0;
+    }
+
+    public function current()
+    {
+        return $this->csIt->current();
+    }
+
+    public function getId()
+    {
+        return $this->csIt->getInnerIterator()->getId();
+    }
+
+    public function key()
+    {
+        return $this->key;
+    }
+
+    public function next()
+    {
+        $this->key++;
+        try {
+            $this->csIt->next();
+            if ($this->valid()) {
+                $this->extractResumeToken($this->csIt->current());
+            }
+        } catch (RuntimeException $e) {
+            $this->resume();
+        }
+    }
+
+    public function resume()
+    {
+        $newChangeStream = call_user_func($this->resumeCallable, $this->resumeToken);
+        $this->csIt = $newChangeStream->csIt;
+        $this->csIt->rewind();
+    }
+
+    public function rewind()
+    {
+        $this->csIt->rewind();
+    }
+
+    public function valid()
+    {
+        return $this->csIt->key() !== null;
```

Comment:
Looking at [`IteratorIterator::key()`](https://github.com/php/php-src/blob/PHP-7.1.11/ext/spl/spl_iterators.c#L1774), I believe this is equivalent to returning [`IteratorIterator::valid()`](https://github.com/php/php-src/blob/PHP-7.1.11/ext/spl/spl_iterators.c#L1753). If you think that's more readable, you can swap to returning `$this->csIt->valid()`, but this should work perfectly fine as-is.

Both of them depend on [`spl_dual_it_fetch()`](https://github.com/php/php-src/blob/PHP-7.1.11/ext/spl/spl_iterators.c#L1693) to populate the current data and key from the inner iterator (our Cursor). Looking at [Cursor's key handler](https://github.com/mongodb/mongo-php-driver/blob/1.3.2/src/MongoDB/Cursor.c#L69) again, I realize I misspoke earlier when I implied that it could return `null` if there was no current element. It actually always returns an integer. That's not technically a problem since we don't implement the userland Iterator interface. PHP's IteratorIterator class and its `key()` and `valid()` methods still function properly because our [current_data handler](https://github.com/mongodb/mongo-php-driver/blob/1.3.2/src/MongoDB/Cursor.c#L77) _does_ return an undefined value if there is no current element. This is because we always free/unset the current element before advancing with the next handler. If libmongoc's cursor doesn't have a next element for us, the Cursor's current element is left unset.